scaaml/capture/input_generators/attack_point_iterator.py

```diff
@@ -0,0 +1,89 @@
+# Copyright 2024 Google LLC
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     https://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""
+An Iterator that iterates through attack points 
+and can be used with config files.
+"""
+
+from abc import ABC, abstractmethod
+from typing import List
+
+
+class AttackPointIterator:
+    """Attack point iterator class that iterates with different configs."""
+
+    def __init__(self, configuration) -> None:
+        """Initialize a new iterator."""
+        self._attack_point_iterator_internal: AttackPointIteratorInternalBase
+        if configuration["operation"] == "constants":
+            constant_iterator = AttackPointIteratorInternalConstants(
+                name=configuration["name"],
+                values=configuration["values"],
+            )
+            self._attack_point_iterator_internal = constant_iterator
+        else:
+            raise ValueError(f"{configuration['operation']} is not supported")
+
+    def __len__(self) -> int:
+        """Return the number of iterated elements.
+        """
+        return len(self._attack_point_iterator_internal)
+
+    def __iter__(self):
+        """Start iterating."""
+        return iter(self._attack_point_iterator_internal)
+
+    def __next__(self):
+        """Next iterated element."""
+        return next(self._attack_point_iterator_internal)
+
+
+class AttackPointIteratorInternalBase(ABC):
+    "Attack point iterator abstract class."
+
+    @abstractmethod
+    def __len__(self) -> int:
+        """Return the number of iterated elements.
+        """
+
+    @abstractmethod
+    def __iter__(self):
+        """Start iterating."""
+
+    @abstractmethod
+    def __next__(self):
+        """Next iterated element."""
+
+
+class AttackPointIteratorInternalConstants(AttackPointIteratorInternalBase):
+    """Attack point iterator class that iterates over a constant."""
+
+    def __init__(self, name: str, values: List[List[int]]) -> None:
+        """Initialize the constants to iterate."""
+        self._name = name
+        self._values = values
+        self._index = 0
+
+    def __len__(self) -> int:
```

Comment:
NIT `__len__` is not unit-tested. I know it is very easy and correct in this case, but for other iterators it is going to be a little harder.

Also the unit-tests have always values of length 2. To find edge cases I would recommend testing length 0, 1, 2, 3. Again you can easily add the following code (I believe leaving the unit-test for docstring is a good idea):

```
def attack_point_iterator_constants(values):
    input = {"operation": "constants", "name": "key", "values": values}
    output = []
    for constant in AttackPointIterator(input):
        output.append(constant[input["name"]])
    assert output == values

def test_attack_point_iterator_constant_lengths():
    for l in range(4):
        values = np.random.randint(256, size=(l, 17))
        attack_point_iterator_constants(values)
```

Similarly testing the call to `len` and testing the number of actual iterations (similar to the repeated iteration tests).